src/middlewared/middlewared/plugins/cloud_sync.py

```diff
@@ -0,0 +1,522 @@
+        return await self.middleware.call('datastore.query', 'system.cloudcredentials', [('id', '=', credentials_id)],
+                                          {'get': True})
+
+    @private
+    async def _validate(self, verrors, name, data):
+        if data['encryption']:
+            if not data['encryption_password']:
+                verrors.add(f'{name}.encryption_password', 'This field is required when encryption is enabled')
+
+            if not data['encryption_salt']:
+                verrors.add(f'{name}.encryption_salt', 'This field is required when encryption is enabled')
+
+        credentials = await self._get_credentials(data['credentials'])
+        provider = REMOTES[credentials["provider"]]
+
+        schema = []
+
+        if provider.buckets:
+            schema.append(Str("bucket"))
+
+        schema.append(Str("folder"))
+
+        schema.extend(provider.task_schema)
+
+        attributes_verrors = validate_attributes(schema, data, additional_attrs=True)
+
+        if not attributes_verrors:
+            await provider.pre_save_task(data, credentials, verrors)
+
+        verrors.add_child(f"{name}.attributes", attributes_verrors)
+
+    @private
+    async def _validate_folder(self, verrors, name, data):
+        if data["direction"] == "PULL":
+            if data["attributes"]["folder"].strip("/"):
+                folder_parent = os.path.normpath(os.path.join(data["attributes"]["folder"].strip("/"), ".."))
+                if folder_parent == ".":
+                    folder_parent = ""
+                folder_basename = os.path.basename(data["attributes"]["folder"].strip("/"))
+                ls = await self.list_directory(dict(
+                    credentials=data["credentials"],
+                    encryption=data["encryption"],
+                    filename_encryption=data["filename_encryption"],
+                    encryption_password=data["encryption_password"],
+                    encryption_salt=data["encryption_salt"],
+                    attributes=dict(data["attributes"], folder=folder_parent)
+                ))
+                for item in ls:
+                    if item["Name"] == folder_basename:
+                        if not item["IsDir"]:
+                            verrors.add(f"{name}.attributes.folder", "This is not a directory")
+                        break
+                else:
+                    verrors.add(f"{name}.attributes.folder", "Directory does not exist")
+
+        if data["direction"] == "PUSH":
+            credentials = await self._get_credentials(data["credentials"])
+
+            provider = REMOTES[credentials["provider"]]
+
+            if provider.readonly:
+                verrors.add(f"{name}.direction", "This remote is read-only")
+
+    @accepts(Dict(
+        'cloud_sync',
+        Str('description'),
+        Str('direction', enum=['PUSH', 'PULL']),
+        Str('transfer_mode', enum=['SYNC', 'COPY', 'MOVE']),
+        Str('path'),
```

Comment:
I think those 3 above should be required=True to avoid a KeyError in validation.